In [102]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [103]:
df = pd.read_csv('melbourne.csv')

In [104]:
df.shape

(23547, 21)

In [105]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         23547 non-null  object 
 1   Address        23547 non-null  object 
 2   Rooms          23547 non-null  int64  
 3   Type           23547 non-null  object 
 4   Price          18396 non-null  float64
 5   Method         23547 non-null  object 
 6   SellerG        23547 non-null  object 
 7   Date           23547 non-null  object 
 8   Distance       23546 non-null  float64
 9   Postcode       23546 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  BuildingArea   10018 non-null  float64
 15  YearBuilt      11540 non-null  float64
 16  CouncilArea    15656 non-null  object 
 17  Lattitude      19243 non-null  float64
 18  Longti

In [106]:
df.head(5)

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

## Identify Missing values

In [107]:
df.isnull() #A boolean function showing all the values in true or false

Suburb  Address  Rooms   Type  Price  Method  SellerG   Date  Distance  \
0       False    False  False  False   True   False    False  False     False   
1       False    False  False  False  False   False    False  False     False   
2       False    False  False  False  False   False    False  False     False   
3       False    False  False  False   True   False    False  False     False   
4       False    False  False  False  False   False    False  False     False   
...       ...      ...    ...    ...    ...     ...      ...    ...       ...   
23542   False    False  False  False   True   False    False  False     False   
23543   False    False  False  False   True   False    False  False     False   
23544   False    False  False  False  False   False    False  False     False   
23545   False    False  False  False  False   False    False  False     False   
23546   False    False  False  False  False   False    False  False     False   

       Postcode  ...  Bathroom    Car  Landsize  BuildingArea  YearBuilt  \
0         False  ...     False  False     False          True       True   
1         False  ...     False  False     False          True       True   
2         False  ...     False  False     False         False      False   
3         False  ...     False  False     False          True       True   
4         False  ...     False  False     False         False      False   
...         ...  ...       ...    ...       ...           ...        ...   
23542     False  ...     False  False     False         False      False   
23543     False  ...      True   True      True          True       True   
23544     False  ...     False  False      True          True       True   
23545     False  ...     False  False     False         False      False   
23546     False  ...     False  False      True         False      False   

       CouncilArea  Lattitude  Longtitude  Regionname  Propertycount  
0            False      False       False       False          False  
1            False      False       False       False          False  
2            False      False       False       False          False  
3            False      False       False       False          False  
4            False      False       False       False          False  
...            ...        ...         ...         ...            ...  
23542         True      False       False       False          False  
23543         True      False       False       False          False  
23544         True      False       False       False          False  
23545         True      False       False       False          False  
23546         True      False       False       False          False  

[23547 rows x 21 columns]

In [108]:
df.isnull().any() #Returns the boolean presence 
#any() return True when atleast one value has null value
#here axis can be set with values axis=0 means column which is the default value

Suburb           False
Address          False
Rooms            False
Type             False
Price             True
Method           False
SellerG          False
Date             False
Distance          True
Postcode          True
Bedroom2          True
Bathroom          True
Car               True
Landsize          True
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude         True
Longtitude        True
Regionname        True
Propertycount     True
dtype: bool

In [109]:
#Summing the missing values in column wise
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

### we can see that some columns have very large missing values.

In [110]:
#all() returns True when all values in the column/row i.e., axis=0/1 has all the values null
df.isnull().all()

Suburb           False
Address          False
Rooms            False
Type             False
Price            False
Method           False
SellerG          False
Date             False
Distance         False
Postcode         False
Bedroom2         False
Bathroom         False
Car              False
Landsize         False
BuildingArea     False
YearBuilt        False
CouncilArea      False
Lattitude        False
Longtitude       False
Regionname       False
Propertycount    False
dtype: bool

In [111]:
#sum of missing values in each columns/rows by setting axis=0/1 respectively
df.isnull().sum(axis=0)

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

## Handling Missing Data

In [112]:
#There are broadly three ways to treat missing values:
#1. Do nothing
#2. Delete the missing values
#3. Impute: Using Simple Statistics - replace the missing values by another value, commonly the mean, median, mode etc.
#   Impute: By Prediction - use models such as k-NN, SVM etc to predict and impute missing values.
#Impute means prediction and giving an approximate data which means changing the actual dataset.

In [113]:
#Let's find the missing values in terms of percentage for columns
def findMissingPercent():
    return round(df.isnull().sum()/len(df.index)*100,2)
    
findMissingPercent()

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

In [157]:
#Deleting missing values by considering the missing values percentage greater than 50% after considering
#if they determine the dependent variable's outcome which in this case is the price
#Removing three columns and note during dropping we give the axis=1 for columns
df = df.drop(['YearBuilt','CouncilArea'],axis=1)
df = df.drop('BuildingArea',axis=1)

In [115]:
#Let's find the missing values in terms of percentage for columns
round(df.isnull().sum()/len(df.index)*100,2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

In [116]:
#Now, instead of removing the columns for which the missing value is more than 20%, we can remove the rows
#which has missing values for more than 5 columns
df[df.isnull().sum(axis=1)>5].head()

Suburb              Address  Rooms Type      Price Method  \
15  Abbotsford     217 Langridge St      3    h  1000000.0      S   
16  Abbotsford      18a Mollison St      2    t   745000.0      S   
19  Abbotsford  403/609 Victoria St      2    u   542000.0      S   
21  Abbotsford    25/84 Trenerry Cr      2    u   760000.0     SP   
22  Abbotsford    106/119 Turner St      1    u   481000.0     SP   

         SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
15        Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
16        Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
19        Dingle  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
21        Biggin  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
22  Purplebricks  10-12-2016       2.5    3067.0       NaN       NaN  NaN   

    Landsize  BuildingArea  Lattitude  Longtitude             Regionname  \
15       NaN           NaN        NaN         NaN  Northern Metropolitan   
16       NaN           NaN        NaN         NaN  Northern Metropolitan   
19       NaN           NaN        NaN         NaN  Northern Metropolitan   
21       NaN           NaN        NaN         NaN  Northern Metropolitan   
22       NaN           NaN        NaN         NaN  Northern Metropolitan   

    Propertycount  
15         4019.0  
16         4019.0  
19         4019.0  
21         4019.0  
22         4019.0

In [117]:
#Counting the values gives us,
len(df[df.isnull().sum(axis=1)>5].index)
#Finding in terms of percentage
round(len(df[df.isnull().sum(axis=1)>5].index)/len(df.index)*100,2)

18.35

In [118]:
#About 18% of rows have data which has more than 5 columns having missing values and those can be removed.
#Hence, we retain only those rows which has missing values not more than 5 columns
df = df[df.isnull().sum(axis=1)<=5]

In [119]:
#Now, lets find the percentage
findMissingPercent()

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.53
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          0.83
Bathroom          0.85
Car               1.59
Landsize          9.45
BuildingArea     47.89
Lattitude         0.14
Longtitude        0.14
Regionname        0.00
Propertycount     0.00
dtype: float64

In [120]:
#We still have the Price column for which the missing values is 20% of data.
#It's a percentage which can be removed.
df = df[~np.isnan(df['Price'])]

In [121]:
#Now, let's find the percentage
findMissingPercent()

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price             0.00
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          1.05
Bathroom          1.07
Car               1.76
Landsize          9.83
BuildingArea     48.55
Lattitude         0.15
Longtitude        0.15
Regionname        0.00
Propertycount     0.00
dtype: float64

In [122]:
#Some columns still have some missing values such as 
df['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

In [123]:
#The mean is skewed here and cannot be used for imputation
#Hence, rows having landsize as null cannot be used and can be safely removed.
df = df[~np.isnan(df.Landsize)]

In [124]:
#The percentage again
findMissingPercent()

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price             0.00
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          0.00
Bathroom          0.01
Car               0.46
Landsize          0.00
BuildingArea     47.52
Lattitude         0.16
Longtitude        0.16
Regionname        0.00
Propertycount     0.00
dtype: float64

In [125]:
#The data looks good now

In [126]:
#lets now go for imputation with rows having very less null values
#Let's go for latitude and longitude
#Notice, loc is used here for indexing the dataframe
df.loc[:,['Lattitude', 'Longtitude']].describe()

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

In [127]:
#The spread of data can be checked by looking to the difference between mean and median, 
#and the variation from 25th to 75th percentile

In [128]:
#we can go for imputation by considering the mean as the difference between mean and median is very low here.
df.loc[np.isnan(df['Lattitude']),['Lattitude']] = df['Lattitude'].mean()
df.loc[np.isnan(df['Longtitude']),['Longtitude']] = df['Longtitude'].mean()

In [129]:
findMissingPercent()

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price             0.00
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          0.00
Bathroom          0.01
Car               0.46
Landsize          0.00
BuildingArea     47.52
Lattitude         0.00
Longtitude        0.00
Regionname        0.00
Propertycount     0.00
dtype: float64

In [130]:
#Now, lets check for Car and Bathroom.
df.loc[:,['Car','Bathroom']].describe()

Car      Bathroom
count  13540.000000  13602.000000
mean       1.610414      1.534921
std        0.962244      0.691834
min        0.000000      0.000000
25%        1.000000      1.000000
50%        2.000000      1.000000
75%        2.000000      2.000000
max       10.000000      8.000000

In [138]:
#The majority i.e., 75th percentile points to 2 which means atleast 75% of row has 2
#We can also find the count.
df['Car'] = df['Car'].astype('category')
df['Car'].value_counts()

2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [139]:
#Similary for Bathrooms
#We can also find the count.
df['Bathroom'] = df['Bathroom'].astype('category')
df['Bathroom'].value_counts()

1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
7.0       2
8.0       2
Name: Bathroom, dtype: int64

In [155]:
#Imputing the car with value having NaN as 2
df.loc[pd.isna(df['Car']),['Car']] = 2
#Imputing the Bathroom with value having NaN as 1
df.loc[pd.isna(df['Bathroom']),['Bathroom']] = 1

In [158]:
findMissingPercent()

Suburb           0.0
Address          0.0
Rooms            0.0
Type             0.0
Price            0.0
Method           0.0
SellerG          0.0
Date             0.0
Distance         0.0
Postcode         0.0
Bedroom2         0.0
Bathroom         0.0
Car              0.0
Landsize         0.0
Lattitude        0.0
Longtitude       0.0
Regionname       0.0
Propertycount    0.0
dtype: float64

In [160]:
#None of the columns have no NaN value anymore
#Checking the shape now
df.shape

(13603, 18)

In [162]:
#Fraction of data remaining now
round(len(df.index)/23547*100,2)

57.77

In [163]:
#Hence, data lost is as below,
100-round(len(df.index)/23547*100,2)

42.23

## Roughly 58% of data is left which is good enough to do regressions.